concent_api/core/views.py

```diff
@@ -302,7 +314,6 @@ def handle_receive_force_get_task_result_upload_for_provider(request, decoded_me
 
     current_time            = int(datetime.datetime.now().timestamp())
     file_transfer_token     = message.concents.FileTransferToken(
-        timestamp                       = current_time,
         token_expiration_deadline       = current_time + settings.TOKEN_EXPIRATION_DEADLINE,
```

Comment:
I would refrain from using assert outside of tests. Mainly because it's disabled in optimized mode (`python -O`) https://docs.python.org/3/library/constants.html#__debug__

Reply:
:+1:

Reply:
@jiivan But that's how `assert` is supposed to be used. It's for development-time sanity checks. For explicitly stating your assumptions in a way that the interpreter can verify and call you out on it if you're wrong. The condition is never supposed to fail and the `-O` flag disables them so that they don't slow down your code in production. They're useless there because the code with broken assumptions should not even pass the testing stage without being repaired.

It's a development tool, not a replacement for the `if ...: raise ...` construct. Not for validating user input or handling errors. It's often used to check function parameters but only as a way to state _"this function does not support parameter values that do not satisfy these conditions and you should write your code in such a way that it never gets called with them"_. Which is of course not appropriate for a public function in an API that's supposed to work with all possible values but is fine in your internal code. If you call the function with bad values anyway, in optimized mode it will just continue and most likely give you wrong results, just like a function without the `assert` would. The difference is that with `assert` if you run into this case during development the interpreter will detect the problem early and signal it by raising `AssertionError` and interrupting the program.

In this case we're checking that our deadline is consistent with the timestamp created by golem-messages. It's always supposed to be and we don't want to add an `if` and check it at runtime. That would be silly. But it's always a good idea to document your assumptions. If we can also get the intepreter to verify them for us for free, that's even better.

Here's a good explanation of the concept: [Assert Statements in Python](https://dbader.org/blog/python-assert-tutorial). This is how it was originally used in the C language. It was a macro that you could disable in optimized code using a compiler flag. In Python the purpose is similar, though I find that many people are not familiar with the intended usage and later the testing frameworks appropriated `AssertionError` for their own purposes. Probably because the name fits really well. I recommend using it a lot because it's an invaluable tool when you're debugging. If you assert your assumptions a lot, your code tends to fail loudly rather than silently do the wrong thing.

Reply:
Still, this means mixing tests with code, right? :)

Reply:
I've never heard anyone refer to checking invariants or preconditions as testing. I think this association comes from the testing frameworks that use assertions too. The only thing in common is that both are comparing stuff and failing if it's not what's expected. But there's no contrived setup, no fixtures, no mocks and we're not running any additional code under test. It's not like, for example, doctests where you run short, one-line test cases and check the output. We're only checking the output of what the program normally does.

To me it's more similar to going through the code in a debugger and checking important values every few steps. Except that it happens automatically because you say what can go wrong in advance. It helps document the code because you're forced to express your implicit assumptions. Often this way you can replace verbose comments that state your assumptions with concise, executable checks at zero performance penalty.